File: 0010-Regular-Expression-Matching/lc0010.py
# Python solution.

class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        ssize, psize = len(s), len(p)
        dp = [[False for i in range(psize+1)] for j in range(ssize+1)]
        dp[ssize][psize] = True

        for i in range(ssize+1)[::-1]:
            for j in range(psize)[::-1]:
                firstMatch = (i < ssize) and (s[i] == p[j] or p[j] == '.')
                if j+1 < psize and p[j+1] == '*':
                    dp[i][j] = dp[i][j+2] or (firstMatch and dp[i+1][j])
                else:
                    dp[i][j] = firstMatch and dp[i+1][j+1]

        return dp[0][0]

File: 0010-Regular-Expression-Matching/lc0010.js
// JavaScript solution.

/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */

var isMatch = function(s, p) {
    let ssize = s.length, psize = p.length;
    let dp = [...Array(ssize+1)].map(e => Array(psize+1).fill(false));
    dp[ssize][psize] = true;

    for (let i = ssize; i >= 0; i--) {
        for (let j = psize-1; j >= 0; j--) {
            let firstMatch = (i < ssize) && (s[i] === p[j] || p[j] === '.');
            if (j+1 < psize && p[j+1] === '*') {
                dp[i][j] = dp[i][j+2] || (firstMatch && dp[i+1][j]);
            } else {
                dp[i][j] = firstMatch && dp[i+1][j+1];
            }
        }
    }
    return dp[0][0];
};
